### PR TITLE
 Refactor zookeeper create client instance #53

### PR DESCRIPTION
**Pull request by paderlol**
Refactor zookeeper create client instance #53

### CODE REVIEW

File: nacossync-worker/src/main/java/com/alibaba/nacossync/extension/holder/AbstractServerHolder.java [hunk 1]
```diff
@@ -35,16 +37,35 @@ public T get(String clusterId, String namespace) throws Exception {
...
+                try {
+                    return createServer(clusterId,skyWalkerCacheServices.getClusterConnectKey(clusterId),
+                            finalNamespace);
+                } catch (Exception e) {
+                    log.error("clusterId={},start server failed", clusterId);
+                    return null;
+                }
+            });
+        } catch (Exception e) {
+            log.error("clusterId={},start server failed", clusterId);
```
Comment (wyp12) [inline nacossync-worker/src/main/java/com/alibaba/nacossync/extension/holder/AbstractServerHolder.java]: 异常栈打印出来

File: nacossync-worker/src/main/java/com/alibaba/nacossync/extension/holder/AbstractServerHolder.java [hunk 2]
```diff
@@ -35,16 +37,35 @@ public T get(String clusterId, String namespace) throws Exception {
...
+            throw new IllegalStateException(e.getMessage(), e);
```
Comment (wyp12) [inline nacossync-worker/src/main/java/com/alibaba/nacossync/extension/holder/AbstractServerHolder.java]: 这段代码看起来有些别扭,catch住异常,又throw出来

**Comment by paderlol** [comment]
这段代码是为了让外面任务中断掉




On 01/16/2019 20:42, Yann wrote:

@wyp12 commented on this pull request.

In nacossync-worker/src/main/java/com/alibaba/nacossync/extension/holder/AbstractServerHolder.java:

>          serviceMapLock.lock();

         try {

-            if (serviceMap.get(key) == null) {

-                serviceMap.put(key, createServer(skyWalkerCacheServices.getClusterConnectKey(clusterId), namespace));

-            }

+            String finalNamespace = namespace;

+            serviceMap.computeIfAbsent(key, clusterKey -> {

+                try {

+                    return createServer(clusterId,skyWalkerCacheServices.getClusterConnectKey(clusterId),

+                            finalNamespace);

+                } catch (Exception e) {

+                    log.error("clusterId={},start server failed", clusterId);

+                    return null;

+                }

+            });

+        } catch (Exception e) {

+            log.error("clusterId={},start server failed", clusterId);



异常栈打印出来

In nacossync-worker/src/main/java/com/alibaba/nacossync/extension/holder/AbstractServerHolder.java:

> -            if (serviceMap.get(key) == null) {

-                serviceMap.put(key, createServer(skyWalkerCacheServices.getClusterConnectKey(clusterId), namespace));

-            }

+            String finalNamespace = namespace;

+            serviceMap.computeIfAbsent(key, clusterKey -> {

+                try {

+                    return createServer(clusterId,skyWalkerCacheServices.getClusterConnectKey(clusterId),

+                            finalNamespace);

+                } catch (Exception e) {

+                    log.error("clusterId={},start server failed", clusterId);

+                    return null;

+                }

+            });

+        } catch (Exception e) {

+            log.error("clusterId={},start server failed", clusterId);

+            throw new IllegalStateException(e.getMessage(), e);



这段代码看起来有些别扭,catch住异常,又throw出来

File: nacossync-worker/src/main/java/com/alibaba/nacossync/extension/holder/AbstractServerHolder.java (hunk 1 above)
Comment (wyp12) [reply nacossync-worker/src/main/java/com/alibaba/nacossync/extension/holder/AbstractServerHolder.java]: log.error("clusterId={},start server failed", clusterId);
===>
  log.error("clusterId={},start server failed", clusterId,**exception**);

File: nacossync-worker/src/main/java/com/alibaba/nacossync/extension/holder/AbstractServerHolder.java (hunk 2 above)
Comment (wyp12) [reply nacossync-worker/src/main/java/com/alibaba/nacossync/extension/holder/AbstractServerHolder.java]: 如果需要中断,是不是不catch就行了 ?

```
try{
xxxxxx
}finally{
            serviceMapLock.unlock();
}
```

Comment (paderlol) [reply nacossync-worker/src/main/java/com/alibaba/nacossync/extension/holder/AbstractServerHolder.java]: 恩 我也是这样想的,因为我异常已经移到函数里面处理掉了

**Pull request merged**